### PR TITLE
{mpi}[GCC/11.3.0] OpenMPI v4.1.4

### PR DESCRIPTION
**Pull request by jfgrimm**
Draft OpenMPI w/ GCC 11.3.0, for latest release candidate. Will update once it is released.

depends on:

- [x] #15456
- [x] #15527

(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/u/UCX/UCX-1.12.1-GCCcore-11.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+easyblock = 'ConfigureMake'
+
+name = 'UCX'
+version = '1.12.1'
```
Comment (boegel) [inline easybuild/easyconfigs/u/UCX/UCX-1.12.1-GCCcore-11.3.0.eb:4]: @jfgrimm I've opened a separate PR for `UCX` in #15451 to already make some progress on this.

UCX is also required as a dependency for `impi`, so it's worth splitting this off imho.
Maybe we should also have a separate PR for `libfabric` + dependencies?

Comment (jfgrimm) [reply easybuild/easyconfigs/u/UCX/UCX-1.12.1-GCCcore-11.3.0.eb]: dropped `UCX` from this PR, and moved `libfabric`, `libevent`, `PMIx` into #15456

**Comment by ocaisa** [comment]
Can we consider adding the patch in https://github.com/easybuilders/easybuild-easyconfigs/pull/14919#issuecomment-1117696051 and
```
configopts = '--with-cuda=internal'
```
?

**Comment by ocaisa** [comment]
@bartoldeman May have another patch to reduce the overhead of CUDA that this might introduce (see [comment](https://github.com/easybuilders/easybuild-easyconfigs/pull/14919#issuecomment-1131743639) from @Micket)

**Comment by Micket** [comment]
I think Bart was looking to get that performance patch merged upstream, so they would likely have some opinions on how to do it and how they like to have their defines. 

But of course, we don't have to care about that when patching all the versions that have already been released. I stole this from the thread (I hope it wasn't secret Bart!)

```patch
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_convertor.c openmpi-4.1.1/opal/datatype/opal_convertor.c
--- openmpi-4.1.1.orig/opal/datatype/opal_convertor.c	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_convertor.c	2022-05-06 11:40:08.698454429 -0400
@@ -41,7 +41,7 @@
 #if OPAL_CUDA_SUPPORT
 #include "opal/datatype/opal_datatype_cuda.h"
 #define MEMCPY_CUDA( DST, SRC, BLENGTH, CONVERTOR ) \
-    CONVERTOR->cbmemcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
+    opal_cuda_memcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
 #endif
 
 static void opal_convertor_construct( opal_convertor_t* convertor )
@@ -51,9 +51,6 @@
     convertor->partial_length = 0;
     convertor->remoteArch     = opal_local_arch;
     convertor->flags          = OPAL_DATATYPE_FLAG_NO_GAPS | CONVERTOR_COMPLETED;
-#if OPAL_CUDA_SUPPORT
-    convertor->cbmemcpy       = &opal_cuda_memcpy;
-#endif
 }
 
 
@@ -694,9 +691,6 @@
         destination->bConverted = source->bConverted;
         destination->stack_pos  = source->stack_pos;
     }
-#if OPAL_CUDA_SUPPORT
-    destination->cbmemcpy   = source->cbmemcpy;
-#endif
     return OPAL_SUCCESS;
 }
 
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_convertor.h openmpi-4.1.1/opal/datatype/opal_convertor.h
--- openmpi-4.1.1.orig/opal/datatype/opal_convertor.h	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_convertor.h	2022-05-06 09:59:06.242836736 -0400
@@ -118,7 +118,6 @@
     dt_stack_t                    static_stack[DT_STATIC_STACK_SIZE];  /**< local stack for small datatypes */
 
 #if OPAL_CUDA_SUPPORT
-    memcpy_fct_t                  cbmemcpy;       /**< memcpy or cuMemcpy */
     void *                        stream;         /**< CUstream for async copy */
 #endif
 };
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_datatype_cuda.c openmpi-4.1.1/opal/datatype/opal_datatype_cuda.c
--- openmpi-4.1.1.orig/opal/datatype/opal_datatype_cuda.c	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_datatype_cuda.c	2022-05-06 10:26:38.659033919 -0400
@@ -48,10 +48,6 @@
         opal_cuda_support_init();
     }
 
-    /* This is needed to handle case where convertor is not fully initialized
-     * like when trying to do a sendi with convertor on the statck */
-    convertor->cbmemcpy = (memcpy_fct_t)&opal_cuda_memcpy;
-
     /* If not enabled, then nothing else to do */
     if (!opal_cuda_enabled) {
         return;
@@ -112,20 +108,17 @@
 }
 
 /*
- * With CUDA enabled, all contiguous copies will pass through this function.
- * Therefore, the first check is to see if the convertor is a GPU buffer.
+ * With CUDA enabled, all contiguous copies will pass through opal_cuda_memcpy which
+ * calls this function. Therefore, that function checks inline to see if the convertor
+ * is a GPU buffer.
  * Note that if there is an error with any of the CUDA calls, the program
  * aborts as there is no recovering.
  */
 
-void *opal_cuda_memcpy(void *dest, const void *src, size_t size, opal_convertor_t* convertor)
+void *opal_cuda_memcpy_gpu(void *dest, const void *src, size_t size, opal_convertor_t* convertor)
 {
     int res;
 
-    if (!(convertor->flags & CONVERTOR_CUDA)) {
-        return memcpy(dest, src, size);
-    }
-
     if (convertor->flags & CONVERTOR_CUDA_ASYNC) {
         res = ftable.gpu_cu_memcpy_async(dest, (void *)src, size, convertor);
     } else {
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_datatype_cuda.h openmpi-4.1.1/opal/datatype/opal_datatype_cuda.h
--- openmpi-4.1.1.orig/opal/datatype/opal_datatype_cuda.h	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_datatype_cuda.h	2022-05-06 10:34:26.120375882 -0400
@@ -10,6 +10,8 @@
 #ifndef _OPAL_DATATYPE_CUDA_H
 #define _OPAL_DATATYPE_CUDA_H
 
+#include <string.h>
+
 /* Structure to hold CUDA support functions that gets filled in when the
  * common cuda code is initialized.  This removes any dependency on <cuda.h>
  * in the opal cuda datatype code. */
@@ -24,10 +26,24 @@
 void mca_cuda_convertor_init(opal_convertor_t* convertor, const void *pUserBuf);
 bool opal_cuda_check_bufs(char *dest, char *src);
 bool opal_cuda_check_one_buf(char *buf, opal_convertor_t *convertor );
-void* opal_cuda_memcpy(void * dest, const void * src, size_t size, opal_convertor_t* convertor);
+void* opal_cuda_memcpy_gpu(void * dest, const void * src, size_t size, opal_convertor_t* convertor);
 void* opal_cuda_memcpy_sync(void * dest, const void * src, size_t size);
 void* opal_cuda_memmove(void * dest, void * src, size_t size);
 void opal_cuda_add_initialization_function(int (*fptr)(opal_common_cuda_function_table_t *));
 void opal_cuda_set_copy_function_async(opal_convertor_t* convertor, void *stream);
 
+/*
+ * With CUDA enabled, all contiguous copies will pass through this function.
+ * Therefore, the first check is to see if the convertor is a GPU buffer.
+ */
+
+static inline void *opal_cuda_memcpy(void *dest, const void *src, size_t size, opal_convertor_t* convertor)
+{
+    if (OPAL_LIKELY(!(convertor->flags & CONVERTOR_CUDA))) {
+        return memcpy(dest, src, size);
+    }
+
+    return opal_cuda_memcpy_gpu(dest, src, size, convertor);
+}
+
 #endif
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_datatype_pack.h openmpi-4.1.1/opal/datatype/opal_datatype_pack.h
--- openmpi-4.1.1.orig/opal/datatype/opal_datatype_pack.h	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_datatype_pack.h	2022-05-06 10:29:01.375528587 -0400
@@ -21,9 +21,10 @@
 
 #if !defined(CHECKSUM) && OPAL_CUDA_SUPPORT
 /* Make use of existing macro to do CUDA style memcpy */
+#include "opal/datatype/opal_datatype_cuda.h"
 #undef MEMCPY_CSUM
 #define MEMCPY_CSUM( DST, SRC, BLENGTH, CONVERTOR ) \
-    CONVERTOR->cbmemcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
+    opal_cuda_memcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
 #endif
 
 /**
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_datatype_unpack.c openmpi-4.1.1/opal/datatype/opal_datatype_unpack.c
--- openmpi-4.1.1.orig/opal/datatype/opal_datatype_unpack.c	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_datatype_unpack.c	2022-05-06 10:16:50.547111046 -0400
@@ -41,6 +41,9 @@
 #include "opal/datatype/opal_datatype_checksum.h"
 #include "opal/datatype/opal_datatype_unpack.h"
 #include "opal/datatype/opal_datatype_prototypes.h"
+#if OPAL_CUDA_SUPPORT
+#include "opal/datatype/opal_datatype_cuda.h"
+#endif
 
 #if defined(CHECKSUM)
 #define opal_unpack_general_function            opal_unpack_general_checksum
@@ -207,7 +210,7 @@
 #if OPAL_CUDA_SUPPORT
     /* In the case where the data is being unpacked from device memory, need to
      * use the special host to device memory copy. */
-    pConvertor->cbmemcpy(saved_data, user_data, data_length, pConvertor );
+    opal_cuda_memcpy(saved_data, user_data, data_length, pConvertor );
 #else
     MEMCPY( saved_data, user_data, data_length );
 #endif
@@ -227,10 +230,10 @@
      * bytes need to be converted back to their original values. */
     {
         char resaved_data[16];
-        pConvertor->cbmemcpy(resaved_data, user_data, data_length, pConvertor );
+        opal_cuda_memcpy(resaved_data, user_data, data_length, pConvertor );
         for(size_t i = 0; i < data_length; i++ ) {
             if( unused_byte == resaved_data[i] )
-                pConvertor->cbmemcpy(&user_data[i], &saved_data[i], 1, pConvertor);
+                opal_cuda_memcpy(&user_data[i], &saved_data[i], 1, pConvertor);
         }
     }
 #else
diff -ur openmpi-4.1.1.orig/opal/datatype/opal_datatype_unpack.h openmpi-4.1.1/opal/datatype/opal_datatype_unpack.h
--- openmpi-4.1.1.orig/opal/datatype/opal_datatype_unpack.h	2021-04-24 13:28:07.000000000 -0400
+++ openmpi-4.1.1/opal/datatype/opal_datatype_unpack.h	2022-05-06 10:29:13.791484625 -0400
@@ -21,9 +21,10 @@
 
 #if !defined(CHECKSUM) && OPAL_CUDA_SUPPORT
 /* Make use of existing macro to do CUDA style memcpy */
+#include "opal/datatype/opal_datatype_cuda.h"
 #undef MEMCPY_CSUM
 #define MEMCPY_CSUM( DST, SRC, BLENGTH, CONVERTOR ) \
-    CONVERTOR->cbmemcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
+    opal_cuda_memcpy( (DST), (SRC), (BLENGTH), (CONVERTOR) )
 #endif
 
 /**
```

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+source_urls = ['https://www.open-mpi.org/software/ompi/v%(version_major_minor)s/downloads']
+sources = [SOURCELOWER_TAR_BZ2]
+patches = ['OpenMPI-4.1.3_build_with_internal_cuda_header.patch']
+checksums = [
+    '46169edd9cf8366608684e5c6a342650b2393d46c9f9a380add8e09f1f8001de',  # openmpi-4.1.4rc1.tar.bz2
+    # OpenMPI-4.1.3_build_with_internal_cuda_header.patch
+    '53bff4fcc92350128b17c29d7288391f87f082f45645c631916b90d2fa5b5116',
+]
+
+builddependencies = [
```
Comment (Micket) [inline easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb:18]: I think we need autotools here.

I was testing this on OpenMPI-4.1.1 and hit
```
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /local/EB/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/config/missing aclocal-1.15 -I config
/local/EB/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/config/missing: line 81: aclocal-1.15: command not found
```

**Review by Micket — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb (hunk 2 above)
Comment (jfgrimm) [reply easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb:18]: yup, added

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name = 'OpenMPI'
+version = '4.1.4rc1'
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb:2]: @jfgrimm OpenMPI 4.1.4 release is there now https://www.open-mpi.org/software/ompi/v4.1/

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb (hunk 3 above)
Comment (jfgrimm) [reply easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb]: :+1:

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+homepage = 'https://www.open-mpi.org/'
+description = """The Open MPI Project is an open source MPI-3 implementation."""
+
+toolchain = {'name': 'GCC', 'version': '11.3.0'}
+
+source_urls = ['https://www.open-mpi.org/software/ompi/v%(version_major_minor)s/downloads']
+sources = [SOURCELOWER_TAR_BZ2]
+patches = ['OpenMPI-4.1.1_build_with_internal_cuda_header.patch']
```
[the remark this answers is not in the document]
Comment (jfgrimm) [reply easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb]: added

**Comment by jfgrimm** [comment]
Test report by @jfgrimm
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
himem01.pri.viking.alces.network - Linux CentOS Linux 7.9.2009, x86_64, Intel(R) Xeon(R) Platinum 8160 CPU @ 2.10GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/fbf16ca89962b4fd38b9dbe3c726bf28 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
bear-pg0105u36b.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/e36284283a9d74b61dc4d340c59bcedb for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
bear-pg0211u08b.bear.cluster - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.8.5
See https://gist.github.com/c35ee6acca6dec90ca1b6c21291755f9 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
bear-pg0306u03a.bear.cluster - Linux RHEL 8.5, POWER, 8335-GTX (power9le), 4 x NVIDIA Tesla V100-SXM2-16GB, 470.57.02, Python 3.6.8
See https://gist.github.com/b3c3401be9550918d34d826d10709fdb for a full test report.

File: easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb (hunk 4 above)
[the remark this answers is not in the document]
Comment (bartoldeman) [reply easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.1.4rc1-GCC-11.3.0.eb]: thanks!

**Comment by bartoldeman** [comment]
@boegelbot please test @ generoso

**Review by bartoldeman — APPROVED** [review]
LGTM

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
fair-mastodon-c6g-2xlarge-0001 - Linux rocky linux 8.5, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/4822d53bb88f5611c9b83b439a56a825 for a full test report.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
node3141.skitty.os - Linux RHEL 8.4, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/ee0b383173a98b1935e8597367a18663 for a full test report.

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ jsc-zen2

**Comment by boegel** [comment]
Going in, thanks @jfgrimm!

**Pull request merged**